drivers: pinctrl: stm32: fix F1 SWJ_CFG application

GeorgeCGV opened the pull request:
Based on `LL` and `HAL` libraries the value of `AFIO_MAPR_SWJ_CFG`  has to be applied to the `MAPR`  upon enabling of a remap. That solved the issue of a not working SPI 1 remap on F103VC and allowed it to be debuggable when the `CONFIG_GPIO_STM32_SWJ_NOJTAG` option is enabled.

The SWJ CFG initialization into `pinctrl` /`pinmux` (*deprecated*) modules.
Kconfig SWJ options were moved into F1 `pinctrl` /`pinmux` DTS.

Fixes #43452
Fixes #43640 

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: gpio: stm32: remove SWJ CFG init

Removed SWJ_CFG initialization. The functionality
is moved into pinctrl / pinmux modules.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: pinctrl: stm32: fix MAPR remap

- Move SWJ_CFG initialization into pinctrl.
- Don't disable the AFIO clock after SWJ_CFG
  initialization.
- Apply '111' to the SWJ_CFG bits upon remap
  application, that fixes the remap usage.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: pinmux: stm32: fix MAPR remap

- Move SWJ_CFG initialization into pinmux.
- Don't disable the AFIO clock after SWJ_CFG
  initialization.
- Apply '111' to the SWJ_CFG bits upon remap
  application, that fixes the remap usage.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

erwango commented on drivers/pinctrl/pinctrl_stm32.c:
```diff
@@ -122,21 +120,37 @@ static int stm32_pins_remap(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt)
+		reg_val = AFIO->MAPR & ~AFIO_MAPR_SWJ_CFG;
+
+		/* set SWJ_CFG bits */
+#if defined(CONFIG_GPIO_STM32_SWJ_NONJTRST)
```
Whole `choice GPIO_STM32_SWJ` should be removed from `drivers/gpio/Kconfig.stm32` and moved to `drivers/pinctrl/Kconfig.stm32` and `drivers/pinmux/Kconfig.stm32`.
This being said, this seems a bit over the kill as pinmux stuff is only there during a deprecation period and will be removed in next DV. Besides, we should consider a clear deprecation message for existing users, which is not easy in that situation.
I'm logging an issue to proceed with that change once pinmux has been removed.

erwango replied:
See https://github.com/zephyrproject-rtos/zephyr/issues/43640

erwango replied:
Thinking to it a bit more, I think we should convert this do device tree, similarly to what was done here: https://github.com/zephyrproject-rtos/zephyr/pull/39106

GeorgeCGV replied:
Agree, the SWJ config within GPIO applies only to the F1. Therefore, it makes no sense to have it there.
Moved into F1 pinctrl dts.

erwango commented on drivers/pinctrl/pinctrl_stm32.c:
```diff
@@ -122,21 +120,37 @@ static int stm32_pins_remap(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt)
 	/* A valid remapping configuration is available */
 	/* Apply remapping before proceeding with pin configuration */
-	LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_AFIO);
+	if (!LL_APB2_GRP1_IsEnabledClock(LL_APB2_GRP1_PERIPH_AFIO)) {
+		LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_AFIO);
+	}
```
We usually don't do that as enabling clock twice has no impact and checking systematically would has footprint impact.

FRASTM commented on drivers/pinmux/pinmux_stm32.c:
```diff
@@ -219,21 +217,37 @@ int stm32_dt_pinctrl_remap(const struct soc_gpio_pinctrl *pinctrl,
-	LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_AFIO);
+	if (!LL_APB2_GRP1_IsEnabledClock(LL_APB2_GRP1_PERIPH_AFIO)) {
+		LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_AFIO);
+	}
```
same as https://github.com/zephyrproject-rtos/zephyr/pull/43605/commits/4ff3e5a819ac13521e7af7877f666934c3900e66#r823483211

GeorgeCGV committed:
dts: bindings: pinctrl: stm32: add SWJ CFG to F1

Adds 'swj-cfg' property to the F1 pinctrl. It is an optional
property that allows to control Serial Wire JTAG
configuration.

Different configurations allow to free certain IO
pins that can be used in remap or standalone.

That replaces previously used Kconfig approach.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: gpio: stm32: remove SWJ from Kconfig

The Serial Wire JTAG configuration is moved to
F1 pinctrl DTS.

The configuration in GPIO didn't apply to the
majority of the STM MCUs except F1.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

erwango commented on drivers/pinctrl/pinctrl_stm32.c:
```diff
@@ -87,6 +87,47 @@ SYS_INIT(stm32_pinmux_init_remap, PRE_KERNEL_1,
+#error Invalid swj-cfg
+#endif
+
+static int stm32_swj_cfg_init(const struct device *dev)
```
stm32f1_swj_cfg_init

erwango commented on drivers/pinctrl/pinctrl_stm32.c:
```diff
@@ -87,6 +87,47 @@ SYS_INIT(stm32_pinmux_init_remap, PRE_KERNEL_1,
 
+#if DT_NODE_HAS_PROP(DT_NODELABEL(pinctrl), swj_cfg)
+
+#if (DT_ENUM_IDX(DT_NODELABEL(pinctrl), swj_cfg) == 0)
+#define PINCTRL_SWJ_CFG_NONJTRST 1
+#elif (DT_ENUM_IDX(DT_NODELABEL(pinctrl), swj_cfg) == 1)
+#define PINCTRL_SWJ_CFG_JTAGDISABLE 1
+#elif (DT_ENUM_IDX(DT_NODELABEL(pinctrl), swj_cfg) == 2)
+#define PINCTRL_SWJ_CFG_DISABLE 1
+#else
+#error Invalid swj-cfg
+#endif
```
Not blocking: we could use directly `(DT_ENUM_IDX(DT_NODELABEL(pinctrl), swj_cfg) == 0)` in the function below.

erwango reviewed (changes requested):
Some minor comments, otherwise lgtm

erwango commented on dts/bindings/pinctrl/st,stm32f1-pinctrl.yaml:
```diff
@@ -11,6 +11,31 @@ properties:
+
+          If absent, then Full SWJ (JTAG-DP + SW-DP) is used (reset state).
+        enum:
+          - "no-njtrst"
```
I wonder if we should add "enable" as a possible value. This is not strictly required, but this may add in clarity.
Not blocking, I'm just opening the discussion.

GeorgeCGV replied:
Added `full` as the first value and made it `default`. 
Used `full` instead of `enable` as it refers to the `swj_cfg` " Full SWJ (JTAG-DP + SW-DP)".

GeorgeCGV commented on dts/bindings/pinctrl/st,stm32f1-pinctrl.yaml:
```diff
@@ -11,6 +11,31 @@ properties:
+          +-----------------+-----------------------------------------------+
+          | disable         | JTAG-DP Disabled and SW-DP Disabled           |
+          |                 | Releases: PA13 PA14 PA15 PB3 PB4.             |
+          +-----------------+-----------------------------------------------+
```
Changed to the the list approach

erwango commented on drivers/pinmux/pinmux_stm32.c:
```diff
@@ -184,6 +184,42 @@ int stm32_dt_pinctrl_configure(const struct soc_gpio_pinctrl *pinctrl,
+	/* reset state is '000' (Full SWJ+JTAG) */
+	/* only one of the 3 bits can be set */
+#if (DT_ENUM_IDX(DT_NODELABEL(pinctrl), swj_cfg) == 1)
+	/* 001 */
```
```suggestion
	/* 001: Full SWJ (JTAG-DP + SW-DP) but without NJTRST */
```

erwango commented on drivers/pinctrl/pinctrl_stm32.c:
```diff
@@ -87,6 +87,41 @@ SYS_INIT(stm32_pinmux_init_remap, PRE_KERNEL_1,
+	/* reset state is '000' (Full SWJ+JTAG) */
+	/* only one of the 3 bits can be set */
+#if (DT_ENUM_IDX(DT_NODELABEL(pinctrl), swj_cfg) == 1)
+	/* 001 */
```
```suggestion
	/* 001: Full SWJ (JTAG-DP + SW-DP) but without NJTRST */
```

erwango commented on drivers/pinctrl/pinctrl_stm32.c:
```diff
@@ -87,6 +87,41 @@ SYS_INIT(stm32_pinmux_init_remap, PRE_KERNEL_1,
+	/* release PB4 */
+	LL_GPIO_AF_Remap_SWJ_NONJTRST();
+#elif (DT_ENUM_IDX(DT_NODELABEL(pinctrl), swj_cfg) == 2)
+	/* 010 */
```
```suggestion
	/* 010: JTAG-DP Disabled and SW-DP Enabled */
```

erwango commented on drivers/gpio/Kconfig.stm32:
```diff
@@ -3,30 +3,8 @@
-
-if GPIO_STM32
-
-choice GPIO_STM32_SWJ
-	prompt "Serial wire JTAG configuration"
-	depends on SOC_SERIES_STM32F1X
-
-config GPIO_STM32_SWJ_ENABLE
-	bool "Full SWJ (JTAG-DP + SW-DP): Reset State"
-
-config GPIO_STM32_SWJ_NONJTRST
-	bool "Full SWJ (JTAG-DP + SW-DP) but without NJTRST"
-
-config GPIO_STM32_SWJ_NOJTAG
-	bool "JTAG-DP Disabled and SW-DP Enabled"
-
-config GPIO_STM32_SWJ_DISABLE
-	bool "JTAG-DP Disabled and SW-DP Disabled"
```
We need to mention this is release note: `doc/releases/release-notes-3.1.rst`
Cf https://github.com/zephyrproject-rtos/zephyr/pull/43062 as exemple

GeorgeCGV committed:
doc: release-notes-3.1: document stm32 f1 swj_cfg

STM32 F1 SWJ_CFG Kconfig options were removed in
favor of pinctrl/pinmux device tree properties.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>
